<a href="https://colab.research.google.com/github/Rossel/DataQuest_Projects/blob/master/Guided_Project_03_Exploring_Ebay_Car_Sales_Data.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Guided Project: Exploring Ebay Car Sales Data

*Practice data cleaning and data explorationg using pandas*

![car image](https://resources.stuff.co.nz/content/dam/images/1/k/6/9/i/l/image.related.StuffLandscapeSixteenByNine.1420x800.1k6a1a.png/1500346348050.jpg)

## 1. Introduction

In this guided project, we'll work with a dataset of used cars from *eBay Kleinanzeigen*, a [classifieds](https://en.wikipedia.org/wiki/Classified_advertising) section of the German eBay website. The data is written in German language, but will be translated where necessary for better understanding.

The dataset was originally scraped and uploaded to [Kaggle](https://www.kaggle.com/orgesleka/used-cars-database/data). DataQuest made the following modifications to the original dataset:

* DataQuest sampled 50,000 data points from the full dataset, to ensure your code runs quickly in our hosted environment
* DataQuest dirtied the dataset a bit to more closely resemble what you would expect from a scraped dataset (the version uploaded to Kaggle was cleaned to be easier to work with)

The dataset can be downloaded [here](https://drive.google.com/file/d/1H8-SUpdMpteA-Qvxn0F1Ad3Ek3z8lU1t/view?usp=sharing). The data dictionary provided with data is as follows:

* `dateCrawled` - When this ad was first crawled. All field-values are taken from this date.
* `name` - Name of the car.
* `seller` - Whether the seller is private or a dealer.
* `offerType` - The type of listing
* `price` - The price on the ad to sell the car.
* `abtest` - Whether the listing is included in an A/B test.
* `vehicleType` - The vehicle Type.
* `yearOfRegistration` - The year in which the car was first registered.
* `gearbox` - The transmission type.
* `powerPS` - The power of the car in PS.
* `model` - The car model name.
* `kilometer` - How many kilometers the car has driven.
* `monthOfRegistration - The month in which the car was first registered.
* `fuelType` - What type of fuel the car uses.
* `brand`- The brand of the car.
* `notRepairedDamage` - If the car has a damage which is not yet repaired.
* `dateCreated` - The date on which the eBay listing was created.
* `nrOfPictures` - The number of pictures in the ad.
* `postalCode` - The postal code for the location of the vehicle.
* `lastSeenOnline` - When the crawler saw this ad last online.

The aim of this project is to clean the data and analyze the included used car listings. We will also become more familiar with some of the unique benefits Jupyter notebook (or Google Colab) provides for pandas.

Let's start by importing the libraries we need and reading the dataset into pandas using Google Colab.



In [15]:
!pip install -U -q PyDrive
from pydrive.auth import GoogleAuth
from pydrive.drive import GoogleDrive
from google.colab import auth
from oauth2client.client import GoogleCredentials

auth.authenticate_user()
gauth = GoogleAuth()
gauth.credentials = GoogleCredentials.get_application_default()
drive = GoogleDrive(gauth)

In [16]:
id = "1H8-SUpdMpteA-Qvxn0F1Ad3Ek3z8lU1t"

In [17]:
downloaded = drive.CreateFile({'id':id}) 
downloaded.GetContentFile('autos.csv')

In [20]:
import pandas as pd
import numpy as np

The dataset could not be read using "UTF-8" or "Windows-1252" encoding, so we used "Latin-1".

In [21]:
autos = pd.read_csv("autos.csv", encoding='Latin-1')

Let's render the first few and last few values of this pandas object, by running the `autos` variable in a separate cell.

In [22]:
autos

dateCrawled  ...             lastSeen
0      2016-03-26 17:47:46  ...  2016-04-06 06:45:54
1      2016-04-04 13:38:56  ...  2016-04-06 14:45:08
2      2016-03-26 18:57:24  ...  2016-04-06 20:15:37
3      2016-03-12 16:58:10  ...  2016-03-15 03:16:28
4      2016-04-01 14:38:50  ...  2016-04-01 14:38:50
...                    ...  ...                  ...
49995  2016-03-27 14:38:19  ...  2016-04-01 13:47:40
49996  2016-03-28 10:50:25  ...  2016-04-02 14:18:02
49997  2016-04-02 14:44:48  ...  2016-04-04 11:47:27
49998  2016-03-08 19:25:42  ...  2016-04-05 16:45:07
49999  2016-03-14 00:42:12  ...  2016-04-06 21:18:48

[50000 rows x 20 columns]

Now we run  the `DataFrame.info()` and `DataFrame.head()` methods to print information about the `autos` dataframe, as well as the first few rows.

In [23]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

In [24]:
autos.head()

dateCrawled  ...             lastSeen
0  2016-03-26 17:47:46  ...  2016-04-06 06:45:54
1  2016-04-04 13:38:56  ...  2016-04-06 14:45:08
2  2016-03-26 18:57:24  ...  2016-04-06 20:15:37
3  2016-03-12 16:58:10  ...  2016-03-15 03:16:28
4  2016-04-01 14:38:50  ...  2016-04-01 14:38:50

[5 rows x 20 columns]

As expected there are 50.000 records across 20 categories. Other observations are:
* Five columns are integers, the others are objects (strings).
* Five columns contain null-values, but none have more than ~20% null values. 
* The column names contain capital letters ("[camel case](https://en.wikipedia.org/wiki/Camel_case)" formatting) instead of Python's preferred "[snakecase](https://en.wikipedia.org/wiki/Snake_case)", which means we cannot simply replace spaces with underscores.

Following this, we will clean column names and drop or fill missing rows with values.

![car image](https://s3.caradvice.com.au/wp-content/uploads/2015/12/BMW-M4-GTS.jpg)

## 2. Cleaning Column Names

Next we will convert the column names from camelcase to snakecase and reword some of the column names based on the data dictionary to be more descriptive.

First we print an array of the existing column names:

In [28]:
print(autos.columns)

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')


Next we will make the following edits to columns names:
* `yearOfRegistration` to `registration_year`
* `monthOfRegistration` to `registration_month`
* `notRepairedDamage` to `unrepaired_damage`
* `dateCreated` to `ad_created`
* The rest of the columnn names from camelcase to snakecase.

We could do quick and clean, by using the following code:
```
autos.columns = [
       'date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'
]
```



However, I would prefer a more structured approach demonstrated below, in case I would work with many more columns, or if future changes would be needed. First, I will make a mapping dictionary, which I then use with the `.rename()` method to change the column names.

In [30]:
column_mapping = {"dateCrawled" : "date_crawled", 
  "name" : "name", 
  "seller" : "seller", 
  "offerType" : "offer_type", 
  "price" : "price", 
  "abtest": "ab_test",
  "vehicleType" :"vehicle_type", 
  "yearOfRegistration" : "registration_year", 
  "gearbox" : "gear_box", 
  "powerPS" : "power_ps", 
  "model" : "model",
  "odometer" : "odometer", 
  "monthOfRegistration" : "registration_month", 
  "fuelType" : "fuel_type", 
  "brand" : "brand",
  "notRepairedDamage" : "unrepaired_damage", 
  "dateCreated" : "ad_created", 
  "nrOfPictures" : "nr_of_pictures",
  "postalCode" : "postal_code",
  "lastSeen": "last_seen"}

In [34]:
autos.rename(column_mapping, axis = 1, inplace = True)

autos.head(1)


date_crawled  ...            last_seen
0  2016-03-26 17:47:46  ...  2016-04-06 06:45:54

[1 rows x 20 columns]

By adopting the standard Python formatting conventions it will be easier for other developers and data scientists to read the notebook.

## 3. Initial Exploration and Cleaning

Now let's do some basic data exploration to determine what other cleaning tasks need to be done. Initially we will look for: 
- Text columns where all or almost all values are the same. These can often be dropped as they don't have useful information for analysis. 
- Examples of numeric data stored as text which can be cleaned and converted.



First, let's take a look at the descriptive statistics for all columns. By entering `include='all'` we will get both categorical and numeric columns:

In [42]:
autos.describe(include='all')

date_crawled         name  ...   postal_code            last_seen
count                 50000        50000  ...  50000.000000                50000
unique                48213        38754  ...           NaN                39481
top     2016-03-11 22:38:16  Ford_Fiesta  ...           NaN  2016-04-07 06:17:27
freq                      3           78  ...           NaN                    8
mean                    NaN          NaN  ...  50813.627300                  NaN
std                     NaN          NaN  ...  25779.747957                  NaN
min                     NaN          NaN  ...   1067.000000                  NaN
25%                     NaN          NaN  ...  30451.000000                  NaN
50%                     NaN          NaN  ...  49577.000000                  NaN
75%                     NaN          NaN  ...  71540.000000                  NaN
max                     NaN          NaN  ...  99998.000000                  NaN

[11 rows x 20 columns]



The columns `price` and `odometer` contain both numbers and letters, with 'km' standing for 'kilometers'. These can be added in the column header. Cars with a `price` of `$0` need to be investigated as well.

Some columns have mostly one and the same value for all rows, which make them candidates to be dropped. These columns are: `seller`, `nr_of_pictures` and `offer_type`. 

The columns `date_crawled`, `ad_created`, and `last_seen` will require date formatting. The column `postal_code` can have its numbers after the decimal removed.

Other anomalies are:
- `registration_year` has a minimal value of 1000, and a maximum value of 9999.
- `power_ps` has a maximum of 17700 horse power.
-  `postal_code` has an entry of four digits, where usually 5 digits are used. Also the entry 99998 seems unusual.

We will start with 

In [63]:
#for cat in autos.columns:
#  print(cat.upper())
#  print('\n' * 1)
#  print(autos[cat].value_counts())
#  print('\n' * 1)
#  print(autos[cat].head(20))
#  print('\n' * 4)

DATE_CRAWLED


2016-03-11 22:38:16    3
2016-03-29 23:42:13    3
2016-03-30 17:37:35    3
2016-03-27 22:55:05    3
2016-04-04 16:40:33    3
                      ..
2016-03-11 13:06:26    1
2016-03-12 16:31:16    1
2016-03-12 10:53:15    1
2016-03-07 18:41:20    1
2016-03-23 16:39:20    1
Name: date_crawled, Length: 48213, dtype: int64


0     2016-03-26 17:47:46
1     2016-04-04 13:38:56
2     2016-03-26 18:57:24
3     2016-03-12 16:58:10
4     2016-04-01 14:38:50
5     2016-03-21 13:47:45
6     2016-03-20 17:55:21
7     2016-03-16 18:55:19
8     2016-03-22 16:51:34
9     2016-03-16 13:47:02
10    2016-03-15 01:41:36
11    2016-03-16 18:45:34
12    2016-03-31 19:48:22
13    2016-03-23 10:48:32
14    2016-03-23 11:50:46
15    2016-04-01 12:06:20
16    2016-03-16 14:59:02
17    2016-03-29 11:46:22
18    2016-03-26 19:57:44
19    2016-03-17 13:36:21
Name: date_crawled, dtype: object





NAME


Ford_Fiesta                              78
Volkswagen_Golf_1.4                      75
BMW_31

## 4. Exploring the Odometer and Price Columns

![car image](https://www.motoringresearch.com/wp-content/uploads/2018/12/14_New_Cars_2019.jpg)

Let's continue exploring the data, specifically looking for data that doesn't look right. We'll start by analyzing the `odometer_km` and `price` columns. Here's the steps we will take:

- Analyze the columns using minimum and maximum values and look for any values that look unrealistically high or low (outliers) that we might want to remove.
- We'll use:
 * `Series.unique().shape` to see how many unique values
 * `Series.describe()` to view min/max/median/mean etc
 * `Series.value_counts()`, with some variations:
    * Chained to `.head()` if there are lots of values.
    * Because `Series.value_counts()` returns a series, we can use `Series.sort_index()` with `ascending= True` or `False` to view the highest and lowest values with their counts (can also chain to `head()` here).
  * When removing outliers, we can do `df[(df["col"] > x ) & (df["col"] < y )]`, but it's more readable to use `df[df["col"].between(x,y)]`


## 5. Exploring the date columns

Let's now move on to the date columns and understand the date range the data covers.

There are 5 columns that should represent date values. Some of these columns were created by the crawler, some came from the website itself. We can differentiate by referring to the data dictionary:



```
- `date_crawled`: added by the crawler
- `last_seen`: added by the crawler
- `ad_created`: from the website
- `registration_month`: from the website
- `registration_year`: from the website
```

Right now, the `date_crawled`, `last_seen`, and `ad_created` columns are all identified as string values by pandas. Because these three columns are represented as strings, we need to convert the data into a numerical representation so we can understand it quantitatively. The other two columns are represented as numeric values, so we can use methods like `Series.describe()` to understand the distribution without any extra data processing.

Let's first understand how the values in the three string columns are formatted. These columns all represent full timestamp values, like so:



In [35]:
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

You'll notice that the first 10 characters represent the day (e.g. `2016-03-12`). To understand the date range, we can extract just the date values, use `Series.value_counts()` to generate a distribution, and then sort by the index.

To select the first 10 characters in each column, we can use `Series.str[:10]`:



In [36]:
print(autos['date_crawled'].str[:10])

0        2016-03-26
1        2016-04-04
2        2016-03-26
3        2016-03-12
4        2016-04-01
            ...    
49995    2016-03-27
49996    2016-03-28
49997    2016-04-02
49998    2016-03-08
49999    2016-03-14
Name: date_crawled, Length: 50000, dtype: object


## 6. Dealing with Incorrect Registration Year Data

![car image](https://cdn2.carbuyer.co.uk/sites/carbuyer_d7/files/f-pace-41_3.jpg)
One thing that stands out from the exploration we did in the last screen is that the `registration_year` column contains some odd values:

- The minimum value is `1000`, before cars were invented
- The maximum value is `9999`, many years into the future

Because a car can't be first registered after the listing was seen, any vehicle with a registration year above 2016 is definitely inaccurate. Determining the earliest valid year is more difficult. Realistically, it could be somewhere in the first few decades of the 1900s.

Let's count the number of listings with cars that fall outside the 1900 - 2016 interval and see if it's safe to remove those rows entirely, or if we need more custom logic.

## 7. Exploring Price by Brand

One of the analysis techniques we learned in this course is aggregation. When working with data on cars, it's natural to explore variations across different car brands. We can use aggregation to understand the `brand` column.

In an earlier mission, we explored how to use loops to perform aggregation. Here's what the process looks like:


```
- Identify the unique values we want to aggregate by
- Create an empty dictionary to store our aggregate data
- Loop over the unique values, and for each:
    - Subset the dataframe by the unique values
    - Calculate the mean of whichever column we're interested in
    - Assign the val/mean to the dict as k/v.
```




## 8. Storing Aggregate Data in a DataFrame

![car image](https://s3.india.com/auto/wp-content/uploads/2017/04/Maserati-at-NYIAS-2017-Ghibli-Nerissimo-edition-studio-w-1.jpg)

In previous part, we aggregated across brands to understand mean price. We observed that in the top 6 brands, there's a distinct price gap.

- Audi, BMW and Mercedes Benz are more expensive
- Ford and Opel are less expensive
- Volkswagen is in between

For the top 6 brands, let's use aggregation to understand the average mileage for those cars and if there's any visible link with mean price. While our natural instinct may be to display both aggregated series objects and visually compare them, this has a few limitations:
- it's difficult to compare more than two aggregate series objects if we want to extend to more columns
- we can't compare more than a few rows from each series object
- we can only sort by the index (brand name) of both series objects so we can easily make visual comparisons

Instead, we can combine the data from both series objects into a single dataframe (with a shared index) and display the dataframe directly. To do this, we'll need to learn two pandas methods:

- [pandas series constructor](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.html)
- [pandas dataframe constructor](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.html)

Here's an example of the series constructor that uses the `brand_mean_prices` dictionary:



In [37]:
bmp_series = pd.Series(brand_mean_prices)
print(bmp_series)

NameError: ignored

The keys in the dictionary became the index in the series object. We can then create a single-column dataframe from this series object. We need to use the `columns` parameter when calling the dataframe constructor (which accepts a array-like object) to specify the column name (or the column name will be set to `0` by default):



In [38]:
df = pd.DataFrame(bmp_series, columns=['mean_price'])
df

NameError: ignored